vllm_spyre/compilation_utils.py

```diff
@@ -0,0 +1,228 @@
+    cache and check to make sure that the current vllm config matches,
+    otherwise the cached artifacts cannot be used.
+
+    For encoder models, we do not allow disabling compilation.
+    """
+
+    req_precompiled_decoder_env_var = "VLLM_SPYRE_REQUIRE_PRECOMPILED_DECODERS"
+
+    if not envs_spyre.VLLM_SPYRE_REQUIRE_PRECOMPILED_DECODERS:
+        return
+
+    if not is_decoder:
+        return
+
+    # If this is a decoder model, disable compilation
+    logger.info(
+        "[PRECOMPILED_WARN]-"
+        "Setting %s because %s is a decoder model",
+        DISABLE_COMPILATION_ENV_VAR, vllm_config.model_config.model)
+    os.environ[DISABLE_COMPILATION_ENV_VAR] = "true"
+
+    # If the user has set req_precompiled_decoder_env_var,
+    # then we need to enforce that they setup their cache
+    torch_cache_dir = os.getenv("TORCH_SENDNN_CACHE_DIR", None)
+    torch_cache_enabled = bool(int(os.getenv("TORCH_SENDNN_CACHE_ENABLE",
+                                             "0")))
+
+    if not torch_cache_dir or not torch_cache_enabled or not os.path.isdir(
+            torch_cache_dir):
+        raise ValueError(
+            f"{req_precompiled_decoder_env_var}=1 requires setting"
+            " TORCH_SENDNN_CACHE_DIR to a valid path and setting " \
+            "TORCH_SENDNN_CACHE_ENABLE=1")
+
+    compilation_config_path = Path(
+        torch_cache_dir) / PRE_COMPILE_MODEL_CONFIG_FILENAME
+    compilation_catalog_path = Path(
+        torch_cache_dir) / PRE_COMPILE_MODEL_CATALOG_FILENAME
+
+    if not compilation_catalog_path.exists() and \
+        not compilation_config_path.exists():
+        raise ValueError(
+            f"{req_precompiled_decoder_env_var}=1 was set, but no "
+            f"pre-compiled model config was found in the "
+            f"TORCH_SENDNN_CACHE_DIR: {str(compilation_config_path)} or"
+            f"{str(compilation_catalog_path)} does not exist")
+
+    if not compilation_catalog_path.is_file() and \
+        not compilation_config_path.is_file():
+        raise ValueError(
+            "{req_precompiled_decoder_env_var}=1 was set, but the "
+            "pre-compiled model config is not a file")
+
+    matching_config = None
+
+    # Note: In below implementation we don't tell user exactly what's wrong
+    # but we do "warn" them about mismatch and provide the list of supported
+    # configuration along with what they have given us.
+    if compilation_catalog_path.is_file():
+        with open(compilation_catalog_path) as f:
+            try:
+                pre_compile_catalog = json.load(f)
+            except json.JSONDecodeError as e:
+                raise ValueError(
+                    f"Precompiled catalog {str(compilation_catalog_path)}"
+                    " is not a valid JSON file") from e
+        match_result = match_from_pre_compile_catalog(pre_compile_catalog,
+                                                      vllm_config)
```

Comment:
In a follow up PR, it would be better to surface all potential matches from the catalog. If there is no exact match on the model name, we could still check for "any" matches for certain configs (eg. vllm_spyre_version).